internal/controller/instaslice_controller.go

```diff
@@ -586,20 +617,73 @@ func (r *InstasliceReconciler) podMapFunc(ctx context.Context, obj client.Object
 	return requests
 }
 
-// SetupWithManager sets up the controller with the Manager.
+// Initialize Prometheus-compatible profiles metrics when the controller starts
+// Adds a background goroutine that waits for Instaslice objects.
+// Proceeds to setupWithManager(mgr) to start the reconciler
+// Does not block the controller from reconciling
+// UpdateCompatibleProfilesMetrics only updates Prometheus metrics,in-memory and do not persist in etcd
 func (r *InstasliceReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	restConfig := mgr.GetConfig()
-
 	var err error
 	r.kubeClient, err = kubernetes.NewForConfig(restConfig)
 	if err != nil {
 		return err
 	}
+	mgrAddErr := mgr.Add(manager.RunnableFunc(func(ctx context.Context) error {
+		log := logr.FromContext(ctx)
+		<-mgr.Elected() // Wait for leader election before executing
+		// Retry mechanism to wait for Instaslice objects
+		var instasliceList inferencev1alpha1.InstasliceList
+		retryErr := wait.PollUntilContextTimeout(ctx, 2*time.Second, 60*time.Second, true, func(ctx context.Context) (bool, error) {
+			if err := r.List(ctx, &instasliceList); err != nil {
+				log.Error(err, "Failed to list Instaslice objects, retrying...")
+				return false, nil
+			}
+			if len(instasliceList.Items) > 0 {
+				log.Info("Instaslice objects found", "count", len(instasliceList.Items))
+				return true, nil
+			}
+			log.Info("No Instaslice objects found, waiting...")
+			return false, nil
+		})
+		if retryErr != nil {
+			log.Error(retryErr, "Failed to fetch Instaslice objects after retries")
+			return nil // Do not block the controller from running, meaning reconciler starts in parallel
+		}
+		// Iterate over Instaslices and update Prometheus metrics
+		for _, instaslice := range instasliceList.Items {
```

Comment:
What would happen when the controller crashes?,  should we wait to update all instaslice objects that already have prometheus metrics or wait for system to catch up?

Reply:
If the controller crashes as we are updating metrics dynamically for all objects, metrics will catch up when the controller restarts. This was what I was talking about when we were discussing.